File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/benchmarks/benchmark_runner.dart
import 'package:runanywhere/runanywhere.dart' as sdk;
import 'package:runanywhere_ai/features/benchmarks/benchmark_types.dart';
import 'package:runanywhere_ai/features/benchmarks/llm_benchmark_provider.dart';
import 'package:runanywhere_ai/features/benchmarks/stt_benchmark_provider.dart';
import 'package:runanywhere_ai/features/benchmarks/tts_benchmark_provider.dart';

/// One provider per modality runs deterministic scenarios through the public
/// SDK against a downloaded model (mirrors iOS `BenchmarkScenarioProvider`).
abstract class BenchmarkScenarioProvider {
  BenchmarkCategory get category;
  List<BenchmarkScenario> scenarios();
  Future<BenchmarkMetrics> execute(
    BenchmarkScenario scenario,
    sdk.ModelInfo model,
  );
}

/// Thrown when a run is cancelled by the user (Dart analogue of Swift's
/// `CancellationError`).
class BenchmarkCancelled implements Exception {
  const BenchmarkCancelled();
}

/// Descriptive runner failure (mirrors iOS `BenchmarkRunnerError`).
class BenchmarkRunnerException implements Exception {
  final String message;
  const BenchmarkRunnerException(this.message);

  factory BenchmarkRunnerException.noModelsAvailable(
    List<BenchmarkCategory> skipped,
  ) {
    final names = skipped.map((category) => category.displayName).join(', ');
    return BenchmarkRunnerException(
      'No downloaded models found for: $names. '
      'Download models first from the Models tab.',
    );
  }

  factory BenchmarkRunnerException.noWorkItems(
    List<BenchmarkCategory> skipped,
  ) {
    if (skipped.isEmpty) {
      return const BenchmarkRunnerException(
        'No benchmarks to run. Select at least one downloaded model '
        'and try again.',
      );
    }
    final names = skipped.map((category) => category.displayName).join(', ');
    return BenchmarkRunnerException(
      'No benchmarks to run. Missing on-disk models for: $names. '
      'Download models first from the Models tab.',
    );
  }

  @override
  String toString() => message;
}

/// Pre-flight snapshot of available models per category (mirrors iOS
/// `BenchmarkPreflightResult`).
class BenchmarkPreflightResult {
  final Map<BenchmarkCategory, List<sdk.ModelInfo>> availableCategories;
  final List<BenchmarkCategory> skippedCategories;
  final int totalWorkItems;

  const BenchmarkPreflightResult({
    required this.availableCategories,
    required this.skippedCategories,
    required this.totalWorkItems,
  });
}

/// One scenario/model pair to execute (mirrors iOS `BenchmarkWorkItem`).
class _BenchmarkWorkItem {
  final BenchmarkCategory category;
  final sdk.ModelInfo model;
  final BenchmarkScenario scenario;

  const _BenchmarkWorkItem({
    required this.category,
    required this.model,
    required this.scenario,
  });
}

/// Orchestrates benchmark execution across all providers (mirrors iOS
/// `BenchmarkRunner.swift`).
class BenchmarkRunner {
  final Map<BenchmarkCategory, BenchmarkScenarioProvider> _providers;

  BenchmarkRunner()
      : _providers = {
          for (final provider in <BenchmarkScenarioProvider>[
            LLMBenchmarkProvider(),
            STTBenchmarkProvider(),
            TTSBenchmarkProvider(),
          ])
            provider.category: provider,
        };

  /// Checks which categories have downloaded models before running, so the
  /// UI can inform the user which categories will be skipped.
  Future<BenchmarkPreflightResult> preflight(
    Set<BenchmarkCategory> categories,
  ) async {
    await sdk.RunAnywhere.models.refreshModelRegistry();

    final listResult = await sdk.RunAnywhere.models.list();
    if (!listResult.success) {
      throw const BenchmarkRunnerException(
        'Failed to fetch available models from the registry.',
      );
    }
    final allModels = listResult.models.models;

    final available = <BenchmarkCategory, List<sdk.ModelInfo>>{};
    final skipped = <BenchmarkCategory>[];

    for (final category in BenchmarkCategory.values) {
      if (!categories.contains(category)) continue;
      final provider = _providers[category];
      if (provider == null) {
        skipped.add(category);
        continue;
      }
      final models = downloadedModels(category, allModels);
      if (models.isEmpty) {
        skipped.add(category);
      } else {
        available[category] = models;
      }
    }

    var totalItems = 0;
    available.forEach((category, models) {
      final scenarioCount = _providers[category]?.scenarios().length ?? 0;
      totalItems += models.length * scenarioCount;
    });

    return BenchmarkPreflightResult(
      availableCategories: available,
      skippedCategories: skipped,
      totalWorkItems: totalItems,
    );
  }

  /// Run all scenarios for the selected categories/models sequentially.
  ///
  /// [isCancelled] is polled between work items (Dart has no structured task
  /// cancellation); a cancelled run throws [BenchmarkCancelled].
  Future<BenchmarkRunOutput> runBenchmarks({
    required Set<BenchmarkCategory> categories,
    Set<String>? modelIds,
    required void Function(BenchmarkProgressUpdate update) onProgress,
    bool Function()? isCancelled,
  }) async {
    final preflightResult = await preflight(categories);

    if (preflightResult.availableCategories.isEmpty ||
        preflightResult.totalWorkItems == 0) {
      throw BenchmarkRunnerException.noModelsAvailable(
        preflightResult.skippedCategories,
      );
    }

    final workItems = _buildWorkItems(
      categories: categories,
      modelIds: modelIds,
      preflightResult: preflightResult,
    );

    if (workItems.isEmpty) {
      throw BenchmarkRunnerException.noWorkItems(
        preflightResult.skippedCategories,
      );
    }

    final total = workItems.length;
    final results = <BenchmarkResult>[];

    for (var index = 0; index < workItems.length; index++) {
      if (isCancelled?.call() ?? false) {
        throw const BenchmarkCancelled();
      }
      final item = workItems[index];

      onProgress(BenchmarkProgressUpdate(
        completedCount: index,
        totalCount: total,
        currentScenario: item.scenario.name,
        currentModel: item.model.name,
      ));

      final provider = _providers[item.category];
      if (provider == null) continue;

      BenchmarkMetrics metrics;
      try {
        metrics = await provider.execute(item.scenario, item.model);
      } on BenchmarkCancelled {
        rethrow;
      } catch (e) {
        metrics = BenchmarkMetrics()
          ..errorMessage =
              '${item.category.displayName} [${item.model.name}]: $e';
      }

      results.add(BenchmarkResult(
        category: item.category,
        scenario: item.scenario,
        modelInfo: ComponentModelInfo.fromModel(item.model),
        metrics: metrics,
      ));
    }

    onProgress(BenchmarkProgressUpdate(
      completedCount: total,
      totalCount: total,
      currentScenario: 'Done',
      currentModel: '',
    ));

    return BenchmarkRunOutput(
      results: results,
      skippedCategories: preflightResult.skippedCategories,
    );
  }

  /// Expand `(category × model × scenario)` into a flat work-item list.
  List<_BenchmarkWorkItem> _buildWorkItems({
    required Set<BenchmarkCategory> categories,
    required Set<String>? modelIds,
    required BenchmarkPreflightResult preflightResult,
  }) {
    final workItems = <_BenchmarkWorkItem>[];
    for (final category in BenchmarkCategory.values) {
      if (!categories.contains(category)) continue;
      final provider = _providers[category];
      final models = preflightResult.availableCategories[category];
      if (provider == null || models == null) continue;

      final filteredModels = (modelIds == null || modelIds.isEmpty)
          ? models
          : models.where((model) => modelIds.contains(model.id)).toList();

      for (final model in filteredModels) {
        for (final scenario in provider.scenarios()) {
          workItems.add(_BenchmarkWorkItem(
            category: category,
            model: model,
            scenario: scenario,
          ));
        }
      }
    }
    return workItems;
  }

  /// Models whose artifacts exist on disk for [category].
  ///
  /// The proto `isDownloaded` flag is not flipped by the C++ registry on
  /// download (see `model_types.dart` `isReadyOnDevice` note), so a non-empty
  /// `localPath` is the on-disk signal — mirrors iOS
  /// `BenchmarkRunner.downloadedModels`.
  static List<sdk.ModelInfo> downloadedModels(
    BenchmarkCategory category,
    List<sdk.ModelInfo> allModels,
  ) {
    return allModels
        .where((model) =>
            model.category == category.modelCategory &&
            !model.builtIn &&
            model.localPath.trim().isNotEmpty)
        .toList();
  }
}

File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmarks/Services/BenchmarkRunner.swift
//
//  BenchmarkRunner.swift
//  RunAnywhereAI
//
//  Orchestrates benchmark execution across all providers.
//

import Foundation
import RunAnywhere

// MARK: - Provider Protocol

protocol BenchmarkScenarioProvider: Sendable {
    var category: BenchmarkCategory { get }
    func scenarios() -> [BenchmarkScenario]
    func execute(
        scenario: BenchmarkScenario,
        model: RAModelInfo
    ) async throws -> BenchmarkMetrics
}

// MARK: - Runner Errors

enum BenchmarkRunnerError: LocalizedError {
    case noModelsAvailable(skippedCategories: [BenchmarkCategory])
    case noWorkItems(skippedCategories: [BenchmarkCategory])
    case fetchModelsFailed(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .noModelsAvailable(let skipped):
            let names = skipped.map(\.displayName).joined(separator: ", ")
            return "No downloaded models found for: \(names). Download models first from the Models tab."
        case .noWorkItems(let skipped):
            if skipped.isEmpty {
                return "No benchmarks to run. Select at least one downloaded model and try again."
            }
            let names = skipped.map(\.displayName).joined(separator: ", ")
            return """
            No benchmarks to run. Missing on-disk models for: \(names). \
            Download models first from the Models tab.
            """
        case .fetchModelsFailed(let error):
            return "Failed to fetch available models: \(error.localizedDescription)"
        }
    }
}

// MARK: - Pre-flight Result

struct BenchmarkPreflightResult: Sendable {
    let availableCategories: [BenchmarkCategory: [RAModelInfo]]
    let skippedCategories: [BenchmarkCategory]
    let totalWorkItems: Int
}

// MARK: - Work Item

/// One scenario/model pair to execute in a single benchmark category.
struct BenchmarkWorkItem: Sendable {
    let category: BenchmarkCategory
    let model: RAModelInfo
    let scenario: BenchmarkScenario
}

// MARK: - Runner

final class BenchmarkRunner {
    private let providers: [BenchmarkCategory: BenchmarkScenarioProvider]

    init() {
        var map: [BenchmarkCategory: BenchmarkScenarioProvider] = [:]
        let all: [BenchmarkScenarioProvider] = [
            LLMBenchmarkProvider(),
            STTBenchmarkProvider(),
            TTSBenchmarkProvider(),
            VLMBenchmarkProvider()
        ]
        for provider in all {
            map[provider.category] = provider
        }
        self.providers = map
    }

    // MARK: - Preflight Check

    /// Checks which categories have downloaded models before running. This lets the UI
    /// inform the user which categories will be skipped.
    func preflight(categories: Set<BenchmarkCategory>) async throws -> BenchmarkPreflightResult {
        await RunAnywhere.refreshModelRegistry()

        let allModels: [RAModelInfo]
        let listResult = await RunAnywhere.listModels()
        guard listResult.success else {
            throw BenchmarkRunnerError.fetchModelsFailed(
                underlying: SDKException(
                    code: .processingFailed,
                    message: listResult.errorMessage.isEmpty ? "model registry" : listResult.errorMessage,
                    category: .internal
                )
            )
        }
        allModels = listResult.models.models

        var available: [BenchmarkCategory: [RAModelInfo]] = [:]
        var skipped: [BenchmarkCategory] = []

        for category in BenchmarkCategory.allCases where categories.contains(category) {
            guard providers[category] != nil else {
                skipped.append(category)
                continue
            }
            let models = Self.downloadedModels(for: category, in: allModels)
            if models.isEmpty {
                skipped.append(category)
            } else {
                available[category] = models
            }
        }

        var totalItems = 0
        for (category, models) in available {
            let scenarioCount = providers[category]?.scenarios().count ?? 0
            totalItems += models.count * scenarioCount
        }

        return BenchmarkPreflightResult(
            availableCategories: available,
            skippedCategories: skipped,
            totalWorkItems: totalItems
        )
    }

    // MARK: - Run

    // swiftlint:disable:next function_body_length
    func runBenchmarks(
        categories: Set<BenchmarkCategory>,
        modelIds: Set<String>? = nil,
        onProgress: @escaping @Sendable (BenchmarkProgressUpdate) -> Void
    ) async throws -> BenchmarkRunOutput {
        let preflight = try await preflight(categories: categories)

        // If nothing to run, throw a descriptive error
        if preflight.availableCategories.isEmpty || preflight.totalWorkItems == 0 {
            throw BenchmarkRunnerError.noModelsAvailable(
                skippedCategories: preflight.skippedCategories
            )
        }

        let workItems = buildWorkItems(
            categories: categories,
            modelIds: modelIds,
            preflight: preflight
        )

        guard !workItems.isEmpty else {
            throw BenchmarkRunnerError.noWorkItems(
                skippedCategories: preflight.skippedCategories
            )
        }

        let total = workItems.count
        var results: [BenchmarkResult] = []

        for (index, item) in workItems.enumerated() {
            try Task.checkCancellation()

            onProgress(BenchmarkProgressUpdate(
                completedCount: index,
                totalCount: total,
                currentScenario: item.scenario.name,
                currentModel: item.model.name
            ))

            let metrics: BenchmarkMetrics
            do {
                guard let provider = providers[item.category] else { continue }
                metrics = try await provider.execute(
                    scenario: item.scenario,
                    model: item.model
                )
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                var errorMetrics = BenchmarkMetrics()
                let prefix = "\(item.category.displayName) [\(item.model.name)]"
                errorMetrics.errorMessage = "\(prefix): \(error.localizedDescription)"
                metrics = errorMetrics
            }

            results.append(BenchmarkResult(
                category: item.category,
                scenario: item.scenario,
                modelInfo: ComponentModelInfo(from: item.model),
                metrics: metrics
            ))
        }

        // Final progress
        onProgress(BenchmarkProgressUpdate(
            completedCount: total,
            totalCount: total,
            currentScenario: "Done",
            currentModel: ""
        ))

        return BenchmarkRunOutput(
            results: results,
            skippedCategories: preflight.skippedCategories
        )
    }

    /// Expand `(category × model × scenario)` into a flat list of work items.
    private func buildWorkItems(
        categories: Set<BenchmarkCategory>,
        modelIds: Set<String>?,
        preflight: BenchmarkPreflightResult
    ) -> [BenchmarkWorkItem] {
        var workItems: [BenchmarkWorkItem] = []
        for category in BenchmarkCategory.allCases where categories.contains(category) {
            guard let provider = providers[category],
                  let models = preflight.availableCategories[category] else { continue }
            let filteredModels: [RAModelInfo]
            if let modelIds, !modelIds.isEmpty {
                filteredModels = models.filter { modelIds.contains($0.id) }
            } else {
                filteredModels = models
            }
            let scenarioList = provider.scenarios()
            for model in filteredModels {
                for scenario in scenarioList {
                    workItems.append(BenchmarkWorkItem(
                        category: category,
                        model: model,
                        scenario: scenario
                    ))
                }
            }
        }
        return workItems
    }

    /// Models whose artifacts exist on disk (registry `isDownloaded` may be stale).
    static func downloadedModels(
        for category: BenchmarkCategory,
        in allModels: [RAModelInfo]
    ) -> [RAModelInfo] {
        allModels.filter { model in
            guard model.category == category.modelCategory, !model.isBuiltIn else { return false }
            if model.isDownloadedOnDisk { return true }
            // Post-download registry may mark downloaded before artifact probe catches up.
            return model.isDownloaded && !model.localPath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }
}
